### ISSUE TITLE
Memory leak in mock provider

### ISSUE THREAD

**Issue by uglyog** [DiUS/pact-jvm#126]
Looks like the mock provider's netty server is leaking memory and/or threads. Symptons include travis build being unable to create new threads.

au.com.dius.pact.consumer.examples.ExampleJavaConsumerPactRuleTest > runTest FAILED
    java.lang.OutOfMemoryError: unable to create new native thread
au.com.dius.pact.consumer.examples.ExampleJavaConsumerPactTest > testPact STANDARD_OUT
    23:37:05.518 [Test worker] WARN  io.netty.channel.AbstractChannel - Force-closing a channel whose registration task was not accepted by an event loop: [id: 0xd0446f8c]
    java.lang.OutOfMemoryError: unable to create new native thread
        at java.lang.Thread.start0(Native Method) [na:1.7.0_76]
        at java.lang.Thread.start(Thread.java:714) [na:1.7.0_76]
        at io.netty.util.concurrent.SingleThreadEventExecutor.startThread(SingleThreadEventExecutor.java:853) ~[netty-common-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:727) ~[netty-common-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe.register(AbstractChannel.java:415) ~[netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.channel.SingleThreadEventLoop.register(SingleThreadEventLoop.java:60) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.channel.SingleThreadEventLoop.register(SingleThreadEventLoop.java:48) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.channel.MultithreadEventLoopGroup.register(MultithreadEventLoopGroup.java:64) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.bootstrap.AbstractBootstrap.initAndRegister(AbstractBootstrap.java:307) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.bootstrap.AbstractBootstrap.doBind(AbstractBootstrap.java:273) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.bootstrap.AbstractBootstrap.bind(AbstractBootstrap.java:269) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.bootstrap.AbstractBootstrap.bind(AbstractBootstrap.java:251) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at unfiltered.netty.Server$$anonfun$2.apply(Server.scala:71) [unfiltered-netty-server_2.11-0.8.2.jar:0.8.2]
        at unfiltered.netty.Server$$anonfun$2.apply(Server.scala:65) [unfiltered-netty-server_2.11-0.8.2.jar:0.8.2]
        at scala.collection.immutable.List.map(List.scala:273) [scala-library-2.11.6.jar:na]
        at unfiltered.netty.Server.start(Server.scala:65) [unfiltered-netty-server_2.11-0.8.2.jar:0.8.2]
        at unfiltered.netty.Server.start(Server.scala:60) [unfiltered-netty-server_2.11-0.8.2.jar:0.8.2]
        at au.com.dius.pact.consumer.UnfilteredMockProvider.start(UnfilteredMockProvider.scala:33) [pact-jvm-consumer_2.11-2.2.3.jar:na]
        at au.com.dius.pact.consumer.StatefulMockProvider.start(MockProvider.scala:37) [pact-jvm-consumer_2.11-2.2.3.jar:na]
        at au.com.dius.pact.consumer.StatefulMockProvider$$anonfun$runAndClose$1.apply(MockProvider.scala:44) [pact-jvm-consumer_2.11-2.2.3.jar:na]
        at au.com.dius.pact.consumer.StatefulMockProvider$$anonfun$runAndClose$1.apply(MockProvider.scala:43) [pact-jvm-consumer_2.11-2.2.3.jar:na]
        at scala.util.Try$.apply(Try.scala:191) [scala-library-2.11.6.jar:na]
        at au.com.dius.pact.consumer.StatefulMockProvider.runAndClose(MockProvider.scala:42) [pact-jvm-consumer_2.11-2.2.3.jar:na]
        at au.com.dius.pact.consumer.ConsumerPactRunner.runAndWritePact(ConsumerPactRunner.scala:28) [pact-jvm-consumer_2.11-2.2.3.jar:na]
        at au.com.dius.pact.model.PactFragment.duringConsumerSpec(PactFragment.scala:13) [pact-jvm-consumer_2.11-2.2.3.jar:2.2.3]
        at au.com.dius.pact.model.PactFragment.runConsumer(PactFragment.scala:21) [pact-jvm-consumer_2.11-2.2.3.jar:2.2.3]
        at au.com.dius.pact.consumer.ConsumerPactTest.testPact(ConsumerPactTest.java:24) [main/:na]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.7.0_76]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_76]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_76]
        at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_76]
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47) [junit-4.11.jar:na]

**Comment by stefan-lz** [comment]
I'm constantly getting this problem and cannot get the gradle build to pass. It's something to do with junit provider rule perhaps?

To reproduce run:

    ./gradlew pact-jvm-consumer-junit:test
    ...
    au.com.dius.pact.consumer.resultstests.MissingRequestConsumerPassesTest > testPact FAILED
        java.lang.AssertionError
           Caused by: java.lang.OutOfMemoryError
    ...
    33 tests completed, 22 failed
    
Yet running this single failing test:

    ./gradlew pact-jvm-consumer-junit:test --tests "*MissingRequestConsumerPassesTest"
    ...
    Totals:
    Ran 1 tests containing 2 assertions.
    0 failures, 0 errors.
    ...
    Success!!!
    :pact-jvm-consumer-junit_2.11:test

Now passes

**Comment by RoadRunner120485** [comment]
Same Problem here:

```
WARN  2015-11-16 14:09:59.593 [main] (Slf4JLogger.java:146) - Force-closing a channel whose registration task was not accepted by an event loop: [id: 0xd0fd68be]
java.lang.OutOfMemoryError: unable to create new native thread
        at java.lang.Thread.start0(Native Method) ~[na:1.8.0_45]
        at java.lang.Thread.start(Thread.java:714) ~[na:1.8.0_45]
        at io.netty.util.concurrent.SingleThreadEventExecutor.startThread(SingleThreadEventExecutor.java:853) ~[netty-common-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:727) ~[netty-common-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe.register(AbstractChannel.java:415) ~[netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.channel.SingleThreadEventLoop.register(SingleThreadEventLoop.java:60) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.channel.SingleThreadEventLoop.register(SingleThreadEventLoop.java:48) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.channel.MultithreadEventLoopGroup.register(MultithreadEventLoopGroup.java:64) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.bootstrap.AbstractBootstrap.initAndRegister(AbstractBootstrap.java:307) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.bootstrap.AbstractBootstrap.doBind(AbstractBootstrap.java:273) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.bootstrap.AbstractBootstrap.bind(AbstractBootstrap.java:269) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at io.netty.bootstrap.AbstractBootstrap.bind(AbstractBootstrap.java:251) [netty-transport-4.0.23.Final.jar:4.0.23.Final]
        at unfiltered.netty.Server$$anonfun$2.apply(Server.scala:71) [unfiltered-netty-server_2.11-0.8.2.jar:0.8.2]
        at unfiltered.netty.Server$$anonfun$2.apply(Server.scala:65) [unfiltered-netty-server_2.11-0.8.2.jar:0.8.2]
        at scala.collection.immutable.List.map(List.scala:273) [scala-library-2.11.7.jar:na]
        at unfiltered.netty.Server.start(Server.scala:65) [unfiltered-netty-server_2.11-0.8.2.jar:0.8.2]
        at unfiltered.netty.Server.start(Server.scala:60) [unfiltered-netty-server_2.11-0.8.2.jar:0.8.2]
        at au.com.dius.pact.consumer.UnfilteredMockProvider.start(UnfilteredMockProvider.scala:33) [pact-jvm-consumer_2.11-3.1.2.jar:3.1.2]
        at au.com.dius.pact.consumer.StatefulMockProvider.start(MockProvider.scala:38) [pact-jvm-consumer_2.11-3.1.2.jar:3.1.2]
        at au.com.dius.pact.consumer.StatefulMockProvider$$anonfun$runAndClose$1.apply(MockProvider.scala:45) [pact-jvm-consumer_2.11-3.1.2.jar:3.1.2]
        at au.com.dius.pact.consumer.StatefulMockProvider$$anonfun$runAndClose$1.apply(MockProvider.scala:44) [pact-jvm-consumer_2.11-3.1.2.jar:3.1.2]
        at scala.util.Try$.apply(Try.scala:192) [scala-library-2.11.7.jar:na]
        at au.com.dius.pact.consumer.StatefulMockProvider.runAndClose(MockProvider.scala:43) [pact-jvm-consumer_2.11-3.1.2.jar:3.1.2]
        at au.com.dius.pact.consumer.ConsumerPactRunner.runAndWritePact(ConsumerPactRunner.scala:28) [pact-jvm-consumer_2.11-3.1.2.jar:3.1.2]
        at au.com.dius.pact.model.PactFragment.duringConsumerSpec(PactFragment.scala:13) [pact-jvm-consumer_2.11-3.1.2.jar:3.1.2]
        at au.com.dius.pact.model.PactFragment.runConsumer(PactFragment.scala:21) [pact-jvm-consumer_2.11-3.1.2.jar:3.1.2]
        at au.com.dius.pact.consumer.groovy.PactBuilder.run(PactBuilder.groovy:216) [pact-jvm-consumer-groovy_2.11-3.1.2.jar:3.1.2]
        at au.com.dius.pact.consumer.groovy.PactBuilder.run(PactBuilder.groovy) [pact-jvm-consumer-groovy_2.11-3.1.2.jar:3.1.2]
        at au.com.dius.pact.consumer.groovy.PactBuilder$run.call(Unknown Source) [pact-jvm-consumer-groovy_2.11-3.1.2.jar:3.1.2]
        at de.xyz.FindCustomerBySubscriptionIdSpec.$spock_feature_1_1(FindByIdRepositoryImplSpec.groovy:115) [test-classes/:na]
        at sun.reflect.GeneratedMethodAccessor41.invoke(Unknown Source) ~[na:na]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_45]
        at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_45]
        at org.spockframework.util.ReflectionUtil.invokeMethod(ReflectionUtil.java:153) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.model.MethodInfo.invoke(MethodInfo.java:84) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:480) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:463) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.runFeatureMethod(BaseSpecRunner.java:405) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.doRunIteration(BaseSpecRunner.java:324) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner$6.invoke(BaseSpecRunner.java:309) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:480) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:463) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.runIteration(BaseSpecRunner.java:288) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.initializeAndRunIteration(BaseSpecRunner.java:278) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.ParameterizedSpecRunner.runIterations(ParameterizedSpecRunner.java:119) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.ParameterizedSpecRunner.runParameterizedFeature(ParameterizedSpecRunner.java:43) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.doRunFeature(BaseSpecRunner.java:262) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner$5.invoke(BaseSpecRunner.java:246) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:480) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:463) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.runFeature(BaseSpecRunner.java:238) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.runFeatures(BaseSpecRunner.java:188) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.doRunSpec(BaseSpecRunner.java:98) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner$1.invoke(BaseSpecRunner.java:84) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:480) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:463) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.runSpec(BaseSpecRunner.java:76) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.BaseSpecRunner.run(BaseSpecRunner.java:67) [spock-core-1.0-groovy-2.4.jar:na]
        at org.spockframework.runtime.Sputnik.run(Sputnik.java:63) [spock-core-1.0-groovy-2.4.jar:na]
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:283) [surefire-junit4-2.18.1.jar:2.18.1]
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:173) [surefire-junit4-2.18.1.jar:2.18.1]
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153) [surefire-junit4-2.18.1.jar:2.18.1]
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:128) [surefire-junit4-2.18.1.jar:2.18.1]
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:203) [surefire-booter-2.18.1.jar:2.18.1]
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:155) [surefire-booter-2.18.1.jar:2.18.1]
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103) [surefire-booter-2.18.1.jar:2.18.1]

```

**Comment by RoadRunner120485** [comment]
We solved our problem: 

a contract between a Provider and a Consumer should be unique in respect of the `given`/`uponRecieving` statements. If you are reusing the same statements within a contract multiple times you should not have different request attributes in `withAttributes` or expected responses in `willRespondWith`. Otherwise it seems that pact tries to merge the states somehow in the resulting pact file, what causes errors.

Maybe this is not correct in theory, but it seems that the merge causes `OutOfMemoryError` on my machine. A colleague of mine got some "Merge Conflict' Errors.

**Comment by uglyog** [comment]
Closing this issue, as a new mock server has been implemented with version 3.4.0.
